cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -124,44 +141,217 @@ def resolve_packages(source_dir: RootedPath) -> list[Package]:
     return packages
 
 
-def create_component_from_package(package: Package, project: Project) -> Component:
-    """Create a SBOM component from a yarn Package."""
-    # if the SBOM generation code grows too much, it may be a good idea to split it into a dedicated
-    # module.
+def create_components(
+    packages: list[Package], project: Project, output_dir: RootedPath
+) -> list[Component]:
+    """Create SBOM components for all the packages parsed from the 'yarn info' output."""
+    package_mapping = {package.parsed_locator: package for package in packages}
+    component_resolver = _ComponentResolver(package_mapping, project, output_dir)
+    return [component_resolver.get_component(package) for package in package_mapping.values()]
 
-    name = _resolve_package_name(package)
 
-    return Component(
-        name=name, version=package.version, purl=_generate_purl_for_package(package, name, project)
-    )
+@dataclass(frozen=True)
+class _ResolvedPackage:
+    """A resolved package.
 
+    Compared to the Package class:
+    - has a name attribute even if the locator doesn't include a reliable name
+      (the name is resolved from the package.json of the package)
+    - has a reliable version (resolved from package.json when necessary)
+    """
 
-def _resolve_package_name(package: Package) -> str:
-    """Resolve the package's real name based on its protocol.
+    locator: Locator
+    name: str
+    version: Optional[str]
 
-    Non-registry deps might have names in their locator string that differ from the one in their
-    package.json file.
+    # TODO: used to make sure purls are unique even for an incomplete implementation
+    #   remove raw_locator when no longer needed
+    raw_locator: str
 
-    Look at the package.json name for every non-registry dependency.
-    """
-    return "placeholder"
+
+class _CouldNotResolve(ValueError):
+    """_ComponentResolver failed to resolve the name or version of a package."""
+
+
+class _ComponentResolver:
```

Comment:
This is a nitpick, but probably worth mentioning it in the early review: I don't quite like the idea of having a class to perform some immediate operation. I think that having attributes gives the idea that we're keeping state for a longer time, but in this case here, it is simply to fulfill all the operations to `get_component`, and then the object can be discarded.

I do like the nesting that the class brings, though, since it creates a clear boundary to show that all methods inside are related to this specific feature. I tried to work around this dilemma [here](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/gomod.py#L244-L289) by using a closure and nested functions, so I wouldn't need to instantiate an object.

Again, I feel it's more of a nitpick, so just sharing my thoughts.

Reply:
> I tried to work around this dilemma [here](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/gomod.py#L244-L289) by using a closure and nested functions, so I wouldn't need to instantiate an object.

Probably not a good idea in this case https://github.com/containerbuildsystem/cachi2/pull/352/commits/d18e686931cc14376d76890cb72b0478ed23eb2f

Reply:
I could un-nest them and pass around the `package_mapping`, `project` and `output_dir` as params where needed if that's better

Reply:
I thought [it](https://github.com/containerbuildsystem/cachi2/commit/d18e686931cc14376d76890cb72b0478ed23eb2f) looked good at first sight, but looking at it again, I think it's easy to forget what `project` or `output_dir` are, since it's a lengthy implementation. That's the only downside I see to it, really.

All in all, I think we're still better off with the class :grin:

Reply:
And since we're nitpicking this piece, why do we have [_scoped_name](https://github.com/containerbuildsystem/cachi2/commit/d18e686931cc14376d76890cb72b0478ed23eb2f) as a standalone function? Do we want to allow uses of it ouside of the _ComponentResolver?

Reply:
Probably not, moved it into the class